nixos/tailscale: interface should be up prior to service finish

sedlund opened the pull request:
as it is, tailscale.service does not finish with usable interface ip's causing services that bind to them to fail with no interface available.

it was not clear to me that i should order services after tailscale-autoconnect.service until reading further into the module.   After a lot of experimentation it also does not wait until after the interface is actually available.

this change allows services that bind to tailscale ip's to schedule `After=tailscale-autoconnect.service` and expect the ip to be configured.

i would like to add comments to the module (in authKeyFile) to suggest that people that bind services to tailscale ip's should order units `.after tailscale-autoconnect`

seeking comment on the approach.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Infinidoge reviewed (commented):
I hate the need for an entire script for this, however from looking over it, it seems pretty solid to me. However it is worth noting that `tailscaled-autoconnect` is only used with `authKeyFile`, so documentation should reflect that the `after` can only guarantee it in that circumstance.

IMO: It may be worth adding this as a requirement for `network-online.target` and direct people to use that where possible, as that is more likely to be useful

sedlund commented:
This has been working well for me ordering my tailscale bound services with `.after` and `.wants` `network-online.target`.

Infinidoge commented on nixos/modules/services/networking/tailscale.nix:
```diff
@@ -78,6 +78,7 @@ in
       description = ''
         A file containing the auth key.
         Tailscale will be automatically started if provided.
+        Services that bind to Tailscale ip's should order After=network-online.target.service
```
```suggestion
        Services that bind to Tailscale IPs should order After=network-online.target
```

Infinidoge reviewed (changes requested):
Small doc fixes, besides that looks good to me!

sedlund commented:
Note: I removed `ActivationPolicy = "manual"` as it is redundant per `systemd.network(5)` when the link is `Unmanaged`.

```
Unmanaged=
           Takes a boolean. When "yes", no attempts are made to bring up or configure matching links, equivalent to when
           there are no matching network files. Defaults to "no".
```

Infinidoge commented on nixos/modules/services/networking/tailscale.nix:
```diff
@@ -78,6 +78,9 @@ in
         Tailscale will be automatically started if provided.
+
+        Services that bind to Tailscale IPs should order using {option}`systemd.services.<name>.after`
+        `networking-online.target`.
```
```suggestion
        `network-online.target`.
```

sedlund commented:
The wantedBy `network-online.target` is not enough.

During a big rebuild / switch - I had a service fail to come up that is .after `network-online.target` before tailscaled-autoconnect was done - which failed deploy.

I think having the services `after = [ "tailscaled-autoconnect.service" ]` is still going to be required.

The problem being if tailscale and a service that requires it get added to the list of units to restart on activation *while the system is booted* `network-online.target` is still active from the original boot.  So the unit needs to know to start after the device is ready, which is provided by `tailscaled-autoconnect`.

Infinidoge commented:
Mmm, lovely edge case. Oh well, so long as it is adequately documented.

sedlund commented:
TIL about `notify` service types.  no need for doing your own timing.  the previous oneshot type would not fail (without manually failing as in the previous commit)

`systemd.service(5)`

```man
TimeoutStartSec=
Configures the time to wait for start-up. If a daemon service does not signal start-up
completion within the configured time, the service will be considered failed and will be
shut down again.

<snip>

Defaults to DefaultTimeoutStartSec= set in the manager, except when Type=oneshot is used,
in which case the timeout is disabled by default (see systemd-system.conf(5)).
```

The case statement catches when in state `Starting`, and if there other things that are happening it will be more clear with the *) default catch

sedlund committed:
nixos/tailscale: tailscaled-autoconnect use Type=notify, wait for Running

Previously the `Starting` state was missed, allowing the service to
complete before the interface was ready, causing services that bind to
Tailscale IPs to fail to start.

Now waits for Tailscale to report `Running` and notifies systemd
accordingly.

Switch the unit to Type=notify to timeout if there is no connection.

Remove `NeedsMachineAuth` gating since it requires client approval in
the console.

sg-qwt commented:
BTW, does it make sense to change `systemd.services.tailscaled-autoconnect.serviceConfig.Type to  "exec";`? I can see several people already done that https://github.com/search?q=tailscaled-autoconnect+exec+NOT+is%3Afork+language%3ANix+&type=code , and it doesn't make sense for tailscaled-autoconnect to hang on people during login I think.

sedlund commented:
@sg-qwt nevermind, I dont think that's it.  It should not wait from what I know now.  

A couple of them from your search mention it is specifically because of some container networking like [this](https://github.com/simonwjackson/mountainous/blob/6d420bd7c137e973380ba35f67e129dc3e3d1c2a/modules/nixos/networking/tailscale/default.nix#L321) .

Without seeing logs / systemd-analyze, I think these changing it to `exec` is just masking some other issue.

SuperSandro2000 reviewed (commented):
otherwise LGTM

Infinidoge reviewed (approved):
Everything looks good to me!

sedlund commented:
rebased over the nixfmt update.

Should I squash?

gshpychka commented:
This leads to `nixos-rebuild switch` failing if tailscale is not started. (if I do `sudo tailscale down` and then `sudo nixos-rebuild switch`, the activation will fail.

```
Oct 01 19:06:21 reaper tailscaled-autoconnect-start[339179]: State = Stopped
Oct 01 19:06:22 reaper tailscaled-autoconnect-start[339179]: State = Stopped
Oct 01 19:06:22 reaper systemd[1]: tailscaled-autoconnect.service: start operation timed out. Terminating.
Oct 01 19:06:22 reaper systemd[1]: tailscaled-autoconnect.service: Failed with result 'timeout'.
Oct 01 19:06:22 reaper systemd[1]: Failed to start tailscaled-autoconnect.service.
```
